float32 alpha
float32 beta